src/main/java/org/javarosa/core/model/instance/ExternalDataInstanceSource.java

```diff
@@ -1,64 +1,86 @@
 
-import com.google.common.collect.ImmutableMultimap;
+import com.google.common.collect.ArrayListMultimap;
 import com.google.common.collect.Multimap;
 
 import org.commcare.core.interfaces.RemoteInstanceFetcher;
 import org.javarosa.core.util.externalizable.DeserializationException;
 import org.javarosa.core.util.externalizable.ExtUtil;
 import org.javarosa.core.util.externalizable.ExtWrapMultiMap;
+import org.javarosa.core.util.externalizable.ExtWrapNullable;
 import org.javarosa.core.util.externalizable.Externalizable;
 import org.javarosa.core.util.externalizable.PrototypeFactory;
-import org.javarosa.xml.util.InvalidStructureException;
-import org.javarosa.xml.util.UnfullfilledRequirementsException;
-import org.xmlpull.v1.XmlPullParserException;
 
 import java.io.DataInputStream;
 import java.io.DataOutputStream;
 import java.io.IOException;
+import java.util.UUID;
+
+import javax.annotation.Nullable;
 
 /**
  * Wrapper class for remote data instances which will materialize the instance data
  * from the source information when needed.
  */
 public class ExternalDataInstanceSource implements InstanceRoot, Externalizable {
 
-    TreeElement root;
-    private String sourceUri;
+    @Nullable
+    private TreeElement root;
     private String instanceId;
-    private Multimap<String, String> requestData;
     private boolean useCaseTemplate;
+    private String reference;
 
-    /**
-     * Externalizable constructor
-     */
-    public ExternalDataInstanceSource() {}
+    @Nullable
+    private String sourceUri;
+    private Multimap<String, String> requestData;
 
-    public ExternalDataInstanceSource(String instanceId, String sourceUri, Multimap<String, String> requestData, boolean useCaseTemplate) {
-        this(instanceId, null, sourceUri, requestData, useCaseTemplate);
-    }
+    @Nullable
+    private UUID storageReferenceId;
 
-    public ExternalDataInstanceSource(String instanceId, TreeElement root, String sourceUri, boolean useCaseTemplate) {
-        this(instanceId, root, sourceUri, ImmutableMultimap.of(), useCaseTemplate);
+    public final static String JR_REMOTE_REFERENCE = "jr://instance/remote";
+
+    public ExternalDataInstanceSource() {
     }
 
-    public ExternalDataInstanceSource(String instanceId, TreeElement root, String sourceUri, Multimap<String, String> requestData, boolean useCaseTemplate) {
+    private ExternalDataInstanceSource(String instanceId, TreeElement root, String reference, boolean useCaseTemplate,
+                                       String sourceUri, Multimap requestData, UUID storageReferenceId) {
         this.instanceId = instanceId;
-        this.sourceUri = sourceUri;
-        this.requestData = requestData;
         this.root = root;
+        this.reference = reference;
         this.useCaseTemplate = useCaseTemplate;
+        this.sourceUri = sourceUri;
+        this.requestData = requestData;
+        this.storageReferenceId = storageReferenceId;
```

Comment:
Maybe validate that one of `sourceUri` and `storageReferenceId` is not null and that both are not supplied?